### PR TITLE
BPTL Dashboard: Add Collection ID search

### PR DESCRIPTION
**Pull request by abhinavjonnada82**
This PR addresses following issue:
-> Allows user to look up specimens based on collection ID
-> Optimized as required & re-used code from research & clinical specimen search
-> Enabled global collection search; lets the user search for collection samples across all sites

https://github.com/episphere/biospecimen/issues/426

Related PR: https://github.com/episphere/connectFaas/pull/364

Checklist:
- [X] Code cleanup
- [X] Check for ES Lint warnings
- [X] Verify test cases
- [X] Check for GIT conflicts
- [ ] Verified unit tests pass with current changes [N/A]
- [ ] Any dependencies or modules  [N/A]
- [X] Attach PoC
- [ ] Add notes

Test Cases Verified with PoC attached: 

1.) Collection ID search page
Screenshot 2023-06-01 at 1 24 59 PM

2.) Look up clinical specimen
Screenshot 2023-06-06 at 3 55 24 PM

3.) Look up research specimen from non nih site
Screenshot 2023-06-06 at 3 54 36 PM

4.) Look up specimen that doesn't exist
Screenshot 2023-06-01 at 12 36 41 PM

5.) Verify if research & clinical dashboard specimen search works & returns only NIH users
Clinical NIH specimen
Screenshot 2023-06-06 at 3 47 05 PM
Research NIH specimen
Screenshot 2023-06-06 at 3 48 57 PM
Lookup specimens from other sites
Screenshot 2023-06-06 at 3 51 26 PM

### CODE REVIEW

File: src/pages/finalize.js [hunk 1]
```diff
@@ -86,17 +88,23 @@ export const finalizeTemplate = (data, specimenData) => {
...
+                        <div class="col-auto">
+                            <button class="btn btn-outline-danger" type="button" id="returnToSpecimenSearch">Return to Search</button>
+                        </div>
                     </div>
+                    ` :`<div class="form-group row">
+                <div class="col-auto">
+                    <button class="btn btn-outline-danger" type="button" data-connect-id="${data.Connect_ID}" id="returnToCollectProcess" data-master-specimen-id="${specimenData['820476880']}">${getWorkflow() === 'research' ? 'Return to Collection Data Entry' : 'Return to Collection Data Entry'}</button>
+                </div>
+                <div class="ml-auto">
+                    <button class="btn btn-outline-warning" data-connect-id="${data.Connect_ID}" data-master-specimen-id="${specimenData['820476880']}" type="button" id="finalizedSaveExit">Exit</button>
```
Comment (Gbarra9) [inline src/pages/finalize.js:102]: Maybe we can `specimenData['820476880']` replace the concept Id and reference the [fieldToConceptIdMapping.js](https://github.com/episphere/biospecimen/blob/dev/src/fieldToConceptIdMapping.js#L72-L73) file

File: src/pages/finalize.js [hunk 2]
```diff
@@ -86,17 +88,23 @@ export const finalizeTemplate = (data, specimenData) => {
...
                 </div>
+                <div class="col-auto">
+                    <button class="btn btn-outline-primary modal-open" data-modal-id="modal1" data-connect-id="${data.Connect_ID}" data-master-specimen-id="${specimenData['820476880']}" type="submit" id="finalizedContinue">Review Complete</button>
```
Comment (Gbarra9) [inline src/pages/finalize.js:105]: We can replace the `820476880` cid with a reference to the field mapping file.

File: src/pages/finalize.js [hunk 3]
```diff
@@ -143,17 +151,21 @@ export const finalizeTemplate = (data, specimenData) => {
...
+    document.getElementById('finalizeForm') && document.getElementById('finalizeForm').addEventListener('submit', (e) => { 
         e.preventDefault();
         const modalId = e.target.getAttribute('data-modal-id');
         openModal(modalId);
     });
 
-    document.getElementById('finalizedConfirmButton').addEventListener('click', async (e) => { 
+    document.getElementById('finalizedConfirmButton') && document.getElementById('finalizedConfirmButton').addEventListener('click', async (e) => { 
         specimenData['410912345'] = 353358909;
         specimenData['556788178'] = new Date().toISOString();
```
Comment (Gbarra9) [inline src/pages/finalize.js:170]: This is more for code readability for the future but these concept Ids can be referenced as well. [Field Mapping Reference](https://github.com/episphere/biospecimen/blob/dev/src/fieldToConceptIdMapping.js#L71C1-L94)

File: src/pages/reports/collectionIdSearch.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        let masterSpecimenId = document.getElementById('masterSpecimenId').value.toUpperCase();
+
+        if(masterSpecimenId.length > masterSpecimenIDRequirement.length) masterSpecimenId = masterSpecimenId.substring(0, masterSpecimenIDRequirement.length);
+
+        if (!masterSpecimenIDRequirement.regExp.test(masterSpecimenId) || masterSpecimenId.length !== masterSpecimenIDRequirement.length) {
+            errorMessage('masterSpecimenId', `Collection ID must be ${masterSpecimenIDRequirement.length} characters long and in CXA123456 format.`, true);
+            return;
+        }
+        showAnimation();
+        const biospecimen = await searchSpecimen(masterSpecimenId);
```
Comment (Gbarra9) [inline src/pages/reports/collectionIdSearch.js:62]: Correct me If I am wrong, but should there be another endpoint that needs to be created or modify an existing one on connectFaas? The `searchSpecimen` is currently site dependent. I can search and get NIH collection IDs but not other sites. Do we need to access all biospecimen documents and filter them based on the collection ID?

biospecimen repo: [searchSpecimen function](https://github.com/episphere/biospecimen/blob/dev/src/shared.js#L804-L813)

connectFass repo: 
[endpoint ](https://github.com/episphere/connectFaas/blob/dev/utils/biospecimen.js#L210-L229)
[searchSpecimen function](https://github.com/episphere/connectFaas/blob/dev/utils/firestore.js#L970-L980)

**Review by Gbarra9 — COMMENTED** [review]
Nice usage of state and reusing the `finalizeTemplate` by adding a `flag` parameter. This may need changes, if the search needs to get all biospecimen collection data and filter them out by specimen id on connectFaas.

File: src/shared.js [hunk 5]
```diff
@@ -801,9 +801,10 @@ export const getBoxesByLocation = async (location) => {
     return res;
 }
 
-export const searchSpecimen = async (masterSpecimenId) => {
+export const searchSpecimen = async (masterSpecimenId, allSitesFlag) => {
     const idToken = await getIdToken();
-    const response = await fetch(`${api}api=searchSpecimen&masterSpecimenId=${masterSpecimenId}`, {
+    const specimenQuery = allSitesFlag === true ? `&masterSpecimenId=${masterSpecimenId}&allSitesFlag=${allSitesFlag}`: `&masterSpecimenId=${masterSpecimenId}`
```
Comment (JoeArmani) [inline src/shared.js:806]: Comment only...Just another way to build the string (factor out the common parts):
```javascript
const specimenQuery = `&masterSpecimenId=${masterSpecimenId}` + (allSitesFlag ? `&allSitesFlag=${allSitesFlag}` : '');
```

**Review by JoeArmani — COMMENTED** [review]
Looks good, I am not familiar with this project like Gene is, but this helped me familiarize myself just a little bit (first time looking at it). I'm also not familiar with the required search functionality Gene mentioned, so these a handful of comments/questions/ideas. Thanks.

-Ternary operators in HTML strings: wrap at highest possible scope. If you can save a div render, wrap the ternary operator outside that div (Gene initially brought this to my attention on something I was doing).
-remove unused imports.
-Should we be updating raw conceptIds to fieldMapping variables as we go across all codebases? I think this is a question for the team, not specific to this PR.

File: src/pages/finalize.js [hunk 6]
```diff
@@ -11,22 +12,31 @@ export const finalizeTemplate = (data, specimenData) => {
...
         <div class="row">
-            <h5>Review Collection Data Entry</h5>
+        ${bptlCollectionFlag === true ? 
+            `` : `<h5>Review Collection Data Entry</h5>`}
         </div>
         </br>
         <div class="row">
             <div class="col">
-                <div class="row"><h5>${data['996038075']}, ${data['399159511']}</h5></div>
+                <div class="row"><h5>${data['996038075'] && data['996038075']}, ${data['399159511'] && data['399159511']}</h5></div>
```
[the remark this answers is not in the document]
Comment (abhinavjonnada82) [reply src/pages/finalize.js:21]: Intresting let me look into nullish coalescing

File: src/pages/reports/collectionIdSearch.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        showAnimation();
+        const biospecimen = await searchSpecimen(masterSpecimenId, true);
+        if (biospecimen.code !== 200) {
+            hideAnimation();
+            showNotifications({ title: 'Not found', body: 'Specimen not found!' }, true)
+            return
+        }
+        const biospecimenData = biospecimen.data;
+        let query = `connectId=${parseInt(biospecimenData.Connect_ID)}&allSiteSearch=${true}`;
+        const response = await findParticipant(query);
```
[the remark this answers is not in the document]
Comment (abhinavjonnada82) [reply src/pages/reports/collectionIdSearch.js:70]: done.

File: src/pages/finalize.js [hunk 8]
```diff
@@ -11,22 +12,31 @@ export const finalizeTemplate = (data, specimenData) => {
...
+                </div>` : ``}
                     <div class="row">
                         <div>Collection Phlebotomist Initials: ${specimenData['719427591'] || ''}</div>
                     </div>` 
                     
-                    : ''
+                    : `<div class="row">
+                    ${bptlCollectionFlag === true ? `
+                        <div>Collection Setting: Clinical</div>` : ``}
+                    </div>`
```
[the remark this answers is not in the document]
Comment (abhinavjonnada82) [reply src/pages/finalize.js:37]: Your method doesn't work if you look at the code above.

**Comment by abhinavjonnada82** [comment]
> Looks good, I am not familiar with this project like Gene is, but this helped me familiarize myself just a little bit (first time looking at it). I'm also not familiar with the required search functionality Gene mentioned, so these a handful of comments/questions/ideas. Thanks.
> 
> -Ternary operators in HTML strings: wrap at highest possible scope. If you can save a div render, wrap the ternary operator outside that div (Gene initially brought this to my attention on something I was doing). -remove unused imports. 
I did such ternary operators in HTML strings: wrapped at highest possible scope in other places. Unused imports would be better if removed by the author or just create another issue just for code cleanup
-Should we be updating raw conceptIds to fieldMapping variables as we go across all codebases? I think this is a question for the team, not specific to this PR.
This should be a separate issue on updating raw conceptIds to fieldMapping variables.

File: src/pages/finalize.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoeArmani) [reply src/pages/finalize.js:37]: ahh yeah syntax is off, but I like the concept.

**Review by JoeArmani — APPROVED** [review]
updates look good

**Review by Gbarra9 — APPROVED** [review]
Looks good. My personal preference, but I think it would be helpful for other reviewers to add your related [connectFaas Pull Request](https://github.com/episphere/connectFaas/pull/364#pullrequestreview-1472336869) to the top. I believe you have this [Biospecimen Pull Request](https://github.com/episphere/biospecimen/pull/536) linked in the related [connectFaas Pull Request](https://github.com/episphere/connectFaas/pull/364#pullrequestreview-1472336869).

File: src/pages/finalize.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (we-ai) [reply src/pages/finalize.js:21]: Agree with Joe.

File: src/pages/reports/collectionIdSearch.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        if(masterSpecimenId.length > masterSpecimenIDRequirement.length) masterSpecimenId = masterSpecimenId.substring(0, masterSpecimenIDRequirement.length);
+
+        if (!masterSpecimenIDRequirement.regExp.test(masterSpecimenId) || masterSpecimenId.length !== masterSpecimenIDRequirement.length) {
+            errorMessage('masterSpecimenId', `Collection ID must be ${masterSpecimenIDRequirement.length} characters long and in CXA123456 format.`, true);
+            return;
+        }
+        showAnimation();
+        const biospecimen = await searchSpecimen(masterSpecimenId, true);
+        if (biospecimen.code !== 200) {
```
Comment (we-ai) [inline src/pages/reports/collectionIdSearch.js:63]:
```suggestion
        if (biospecimen.code !== 200 || Object.keys(biospecimen.data).length === 0) {
```
We're updating backend API to return empty data when code 200, if search returns no result.

**Review by we-ai — APPROVED** [review]
Minor change suggested.

**Pull request merged**